crates/router/src/core/user_role/role.rs

```diff
@@ -0,0 +1,180 @@
+    role::{self as role_api},
+    Permission,
+};
+use common_utils::generate_id_with_default_len;
+use diesel_models::role::{RoleNew, RoleUpdate};
+use error_stack::ResultExt;
+
+use crate::{
+    core::errors::{StorageErrorExt, UserErrors, UserResponse},
+    routes::AppState,
+    services::{
+        authentication::UserFromToken,
+        authorization::roles::{self, predefined_roles::PREDEFINED_ROLES},
+        ApplicationResponse,
+    },
+    types::domain::user::RoleName,
+};
+
+pub async fn get_role_from_token(
+    state: AppState,
+    user_from_token: UserFromToken,
+) -> UserResponse<Vec<Permission>> {
+    let role_info = user_from_token
+        .get_role_info_from_db(&state)
+        .await
+        .attach_printable("Invalid role_id in JWT")?;
+
+    let permissions = role_info
+        .get_permissions_set()
+        .into_iter()
+        .map(Into::into)
+        .collect();
+
+    Ok(ApplicationResponse::Json(permissions))
+}
+
+pub async fn create_role(
+    state: AppState,
+    user_from_token: UserFromToken,
+    req: role_api::CreateRoleRequest,
+) -> UserResponse<()> {
+    let now = common_utils::date_time::now();
+    let role_name = RoleName::new(req.role_name)?.get_role_name();
+
+    if req.groups.is_empty() {
+        return Err(UserErrors::InvalidRoleOperation.into())
+            .attach_printable("role groups cannot be empty");
+    }
+
+    state
+        .store
+        .insert_role(RoleNew {
+            role_id: generate_id_with_default_len("role"),
+            role_name,
+            merchant_id: user_from_token.merchant_id,
+            org_id: user_from_token.org_id,
+            groups: req.groups,
+            scope: req.scope,
+            created_by: user_from_token.user_id.clone(),
+            last_modified_by: user_from_token.user_id,
+            created_at: now,
+            last_modified_at: now,
+        })
+        .await
+        .change_context(UserErrors::InternalServerError)?;
+
+    Ok(ApplicationResponse::StatusOk)
+}
+
+pub async fn update_role(
+    state: AppState,
+    user_from_token: UserFromToken,
+    req: role_api::UpdateRoleRequest,
+) -> UserResponse<()> {
+    let now = common_utils::date_time::now();
+
+    let role_name = req
+        .role_name
+        .map(RoleName::new)
+        .transpose()?
+        .map(|x| x.get_role_name());
+
+    if let Some(ref groups) = req.groups {
+        if groups.is_empty() {
+            return Err(UserErrors::InvalidRoleOperation.into())
+                .attach_printable("role groups cannot be empty");
+        }
+    }
+
+    state
+        .store
+        .update_role_by_role_id(
+            &req.role_id,
+            RoleUpdate::UpdateDetails {
+                groups: req.groups,
+                role_name,
+                last_modified_at: now,
+                last_modified_by: user_from_token.user_id,
+            },
+        )
+        .await
+        .change_context(UserErrors::InternalServerError)?;
```

Comment:
We need to check if the role_id given in the request is present in merchant scope or not. For now we can add a find query to check that.